tests/samplers_tests/test_nsgaii.py

```diff
@@ -146,6 +148,134 @@ def constraints_func(trial: FrozenTrial) -> Sequence[float]:
         assert trial.system_attrs[_CONSTRAINTS_KEY] == (trial.number,)
 
 
+@pytest.mark.parametrize("dir1", [StudyDirection.MINIMIZE, StudyDirection.MAXIMIZE])
+@pytest.mark.parametrize("dir2", [StudyDirection.MINIMIZE, StudyDirection.MAXIMIZE])
+@pytest.mark.parametrize(
+    "constraints_list",
+    [
+        [[]],  # empty constraint
+        [[-float("inf")], [-1], [0]],  # 1d constraints
+        [
+            [c1, c2] for c1 in [-float("inf"), -1, 0] for c2 in [-float("inf"), -1, 0]
+        ],  # 2d-constraints
+    ],
+)
+def test_constrained_dominates_feasible_vs_feasible(
+    dir1: StudyDirection, dir2: StudyDirection, constraints_list: List[List[float]]
+) -> None:
+
+    directions = [dir1, dir2]
+
+    # Check all pairs of trials consisting of these values, i.e.,
+    # [-inf, -inf], [-inf, -1], [-inf, 1], [-inf, inf], [-1, -inf], ...
+    values_1d = [-float("inf"), -1, 1, float("inf")]
+
+    # In the following code, we check that the results of _constrained_dominates match
+    # _dominates in all feasible cases.
+    values_list = [[x, y] for x in values_1d for y in values_1d]
+    values_constraints_list = [(vs, cs) for vs in values_list for cs in constraints_list]
+
+    for (values1, constraints1) in values_constraints_list:
+        for (values2, constraints2) in values_constraints_list:
+            t1 = _create_frozen_trial(0, values1, constraints1)
+            t2 = _create_frozen_trial(1, values2, constraints2)
+            assert _constrained_dominates(t1, t2, directions) == _dominates(t1, t2, directions)
```

Comment:
I think that would result in many tests in total. That would slow down the execution, and also flood your terminal with superfluous error messages when there are bugs in the tested function.

Reply:
> That would slow down the execution

It completed in a several seconds in my environment, so I don't think it is a serious drawback.

> I think that would result in many tests in total(4 x 256 x 3 tests). That would slow down the execution, and also flood your terminal with superfluous error messages when there are bugs in the tested function.

The error message can be suppressed by pytests options such as `-x`.

Reply:
I also think that the first code has less duplication and is simpler. What do you think, @HideakiImamura ?

Reply:
Maybe we should write it out in the definition of `values_list`?

Reply:
Thank you for your suggestion. I think twice and the readability issue may caused by the code comment and variable name.

- I expect the comment `Check all pairs of trials consisting of these values ...` be right before `values_list` instead of `values_1d` since the comment explains 2d objective values, not 1d.
- I expect the comment `In the following code ...`  be near assertion, and we can emphasis on `the results of _constrained_dominates match _dominates in all feasible cases`.
- The name of `values_1d` might be confusing since this test case handles both 1d constraints and 2d constraints. Readers might think this test case also handles single objective optimization problem in addition to bi-objective optimization problem.

```suggestion
@pytest.mark.parametrize("dir1", [StudyDirection.MINIMIZE, StudyDirection.MAXIMIZE])
@pytest.mark.parametrize("dir2", [StudyDirection.MINIMIZE, StudyDirection.MAXIMIZE])
@pytest.mark.parametrize(
    "constraints_list",
    [
        [[]],  # empty constraint
        [[-float("inf")], [-1], [0]],  # 1d constraints
        [
            [c1, c2] for c1 in [-float("inf"), -1, 0] for c2 in [-float("inf"), -1, 0]
        ],  # 2d-constraints
    ],
)
def test_constrained_dominates_feasible_vs_feasible(
    dir1: StudyDirection, dir2: StudyDirection, constraints_list: List[List[float]]
) -> None:
    directions = [dir1, dir2]
    # Check all pairs of trials consisting of these values, i.e.,
    # [-inf, -inf], [-inf, -1], [-inf, 1], [-inf, inf], [-1, -inf], ...
    values_list = [[x, y]
                    for x in [-float("inf"), -1, 1, float("inf")]
                    for y in [-float("inf"), -1, 1, float("inf")]]
    values_constraints_list = [(vs, cs) for vs in values_list for cs in constraints_list]

    # The results of _constrained_dominates match _dominates in all feasible cases.
    for (values1, constraints1) in values_constraints_list:
        for (values2, constraints2) in values_constraints_list:
            t1 = _create_frozen_trial(0, values1, constraints1)
            t2 = _create_frozen_trial(1, values2, constraints2)
            assert _constrained_dominates(t1, t2, directions) == _dominates(t1, t2, directions)
```